### ISSUE TITLE
SPI driver transmits wrong values

### ISSUE THREAD

**Issue by gamer-cndg** [OnionIoT/spi-gpio-driver#4]
Using a simple code snippet which sends `0x00` to `0xff` incremental over SPI and a logic analyzer, I can see that the SPI driver transmits wrong values (mostly MSB seems to be wrong / set to 0 in **some** cases). I am doing 8-bit per transfer, 1 byte at a time transmission with standard SPI settings (SPIMODE_0, MSBFIRST)

I created a Google spreadsheet with the tested an measured values: https://docs.google.com/spreadsheets/d/1IKgfG4rZbJNBUeTCGGXWhrnAf6ZNob7rlFdtQVE7zsc/edit?usp=sharing
All values from 0x40 to 0xbf are transmitted wrong (50% of all values)

In this state, the SPI driver cannot be used. When can a fix be expected?

**Comment by 1v3ry** [comment]
+1, same here.
nice to know that i could have spared those hours.

furthermore it seems like it doesn't matter which values you set as `mode`, `modeBits` or whatever. the output looks always the same... i attached a screenshot of the the output for `spi.write([0x41])` with the following setup (i tried different settings in the commented lines, which has no effect to the output):
`spi  = onionSpi.OnionSpi(1, 32766)`
`spi.setVerbosity(1)`
`spi.lsbfirst = False`
`#spi.bitsPerWord = 0`
`spi.sck = 7`
`spi.mosi = 8`
`spi.miso = 9`
`spi.cs = 6`
`spi.speed = 100000`
`spi.modeBits = 0x00`
`#spi.delay = 25`
`# self.spi.csHigh  = 1`
![spi](https://user-images.githubusercontent.com/6094062/42128993-112d735c-7cb8-11e8-8f9e-e9140b10ccca.png)

**Comment by maxgerhardt** [comment]
What firmware version are you on (`oupgrade -c -l`)? Writing wrong SPI values should have been fixed long ago (https://onion.io/2bt-march-27-2018/)

**Comment by 1v3ry** [comment]
well this is funny.
i freshly set the omega2 up via the web-gui (never has been out of the package before) where it said that it's updating the firmware. 

after rebooting i did the following
`root@Omega-5BFD:~# oupgrade`
`> Device Firmware Version: 0.1.10 b160`
`> Checking latest version online...`
`> Repo Firmware Version: 0.1.10 b160`
`> Comparing version numbers`
`> Device firmware is up to date!`
and assumed, that i am now running the latest firmware.

However, i now checked with the `-l` option, et voila:
`root@Omega-5BFD:~# oupgrade -c -l`
`> Device Firmware Version: 0.1.10 b160`
`> Checking latest version online...`
`> Repo Firmware Version: 0.2.0 b188`
`> Comparing version numbers`
`> New firmware version available, need to upgrade device firmware`

Maybe you should at least change the [docs](https://docs.onion.io/omega2-docs/first-time-setup-command-line.html#update-the-omegas-firmware) to mention the `oupgrade -l` command instead of just `oupgrade`. Otherwise i think this issue can be closed, SPI is now sending the bytes it's supposed to.

### TL;DR:
 Update your onion's firmware with `oupgrade -l` via a terminal (you will have to reinstall packages you previously installed). Then you will get the latest firmware and working SPI.

**Comment by maxgerhardt** [comment]
@1v3ry out of interest, are you able to read data? I have an open pending ticket that I'm not able to read any data sent over SPI (sending works) on b188.

**Comment by 1v3ry** [comment]
any news so far?

**Comment by maxgerhardt** [comment]
I'd have to retest this is on the newest firmware, didn't find time yet. Did you test it on the newest version?

**Comment by 1v3ry** [comment]
just found some time to test it. i hope i got the most recent firmware this time:
```
Device Firmware Version: 0.2.0 b195
Checking latest version online...
Repo Firmware Version: 0.1.10 b160
Comparing version numbers
Device firmware is up to date!
```

what i tried again is:

> When i try to
> ```
>  def SetBitMask(self, reg, mask):
>     tmp = self.Read_MFRC522(reg)
>     print tmp
>     print tmp | mask
>     self.Write_MFRC522(reg, tmp | mask)
> ```
> with reg = 0x28 and mask = 0x03
> i get the following output:
> 
> ```
> onion-spi:: Trasferring 0xa8, 2 bytes # the change from 0x28 to 0xa8 is done by the python lib and is **correct and necessary**
>    Received: 0x00, ioctl status: 2
> 64
> 67
> onion-spi:: Trasferring 0x28, 2 bytes
>    Received: 0x00, ioctl status: 2
> ```
> 
> So i'm sending 0xa8 to the reader, which means to read it's register 0x28.
> Next the reader responds with 0x80 (i can see that on my LLA). Now onion-spi verbous mode is saying it doesn't get an answer (aka 0x00), but the tmp variable is 0x40, so the MISO response is read but shifted 1 bit to the right.

According to my LLA the reader is still responding with 0x80 to the read-request at register 0x28.
However the verbous mode of spi is still telling me it read 0x00, no matter what.
The `tmp` variable changes from try to try. i got 0x30, 0x40, 0x60 or 0x70 but the analyzed traffic on the MISO, MOSI, SDA and SCL was the same during all tries.

```
onion-spi:: Trasferring 0xa8, 2 bytes
   Received: 0x00, ioctl status: 2
tmp 48
mask 3
tmp | mask 51

onion-spi:: Trasferring 0xa8, 2 bytes
   Received: 0x00, ioctl status: 2
tmp 64
mask 3
tmp | mask 67

onion-spi:: Trasferring 0xa8, 2 bytes
   Received: 0x00, ioctl status: 2
tmp 96
mask 3
tmp | mask 99

onion-spi:: Trasferring 0xa8, 2 bytes
   Received: 0x00, ioctl status: 2
tmp 112
mask 3
tmp | mask 115
```

**Comment by greenbreakfast** [comment]
For reference: SPI transfer issues on the Omega2 have been fixed since firmware version 0.2.0 

More details here: https://onion.io/2bt-brand-new-os-release/
